Prefill shipping category forms with default values

daningenthron opened the pull request:
#### What? Why?

Closes #4165

Shipping category is not marked as required on the Create Product form. Current configuration allows creation of only one shipping category. This prefills the form field.

#### What should we test?

`/spec/features/admin/products_spec.rb` has been updated to check for the prefilled value instead of adding a value when populating the form.

#### Release notes

This exposed a couple potential issues (possibly only in development environment?):
- Should a user be able to have more than one shipping category saved? There is only an option to edit the default category, not add a second one.
- Attempting to delete the existing shipping category causes an ActiveRecord error.

Changelog Category: Fixed

luisramos0 commented:
if we go ahead with selecting default, it should be encapsulated in a service like DefaultStockLocation, see [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/services/default_stock_location.rb)

This will evolve the approach in existing migration AddShippingCategoryToShippingMethodsAndProducts.

Something like this?

```
# Encapsulates the concept of default shipping category
class DefaultShippingCategory
  NAME = 'Default'.freeze

  def self.create!
    Spree::ShippingCategory.create!(name: NAME)
  end

  def self.find_or_create
    Spree::ShippingCategory.find_or_create_by_name(NAME)
  end
end
```

and then use DefaultShippingCategory.find_or_create in the views

daningenthron commented:
Hi @luisramos0 thanks for clarifying the need for the default! This makes sense looking at #4166 and I agree that it would be better to address both problems with the same fix. I'll add the default service and tie it into both pages.

daningenthron commented:
I believe this sets up the service for use in product and shipping method creation. I adjusted the specs for these pages to use the Defaults as well, and also added it to the seeds file. Let me know what you think!

luisramos0 commented on app/services/default_shipping_category.rb:
```diff
@@ -0,0 +1,13 @@
+# Encapsulates the concept of default stock location in creation of a product or a shipping method.
+
```
this empty line can be removed (I wonder if rubocop is not flagging it)

luisramos0 commented on app/services/default_shipping_category.rb:
```diff
@@ -0,0 +1,13 @@
+  NAME = 'Default'.freeze
+
+  def self.create!
+	  Spree::ShippingCategory.create!(name: NAME)
```
indentation.

luisramos0 commented on spec/services/default_shipping_category_spec.rb:
```diff
@@ -0,0 +1,39 @@
+require 'spec_helper'
+
+describe DefaultShippingCategory do
+	describe '.create!' do
```
indentation needs to be two spaces, not a tab

luisramos0 commented on db/seeds.rb:
```diff
@@ -52,3 +52,4 @@ def set_mail_configuration
 require File.join(File.dirname(__FILE__), 'default', 'users')
 
 DefaultStockLocation.find_or_create
+DefaultShippingCategory.find_or_create
```
there's a new line missing here

luisramos0 commented on npm-debug.log:
```diff
@@ -0,0 +1,27 @@
+10 error missing script: build
+11 error If you need help, you may report this error at:
+11 error     <https://github.com/npm/npm/issues>
+12 verbose exit [ 1, true ]
```
this file must be removed

daningenthron commented:
I think this failed on Semaphore bc of the Ruby version; rebased and resubmitted!

luisramos0 commented:
60 commits... something weird with the rebase.
you can do:
git rebase upstream/master

luisramos0 commented:
ouch :-O I was trying to fix the rebase problem for you and I just overwrote your branch with... nothing. 
To fix this and get back to where this was, you need to force PUSH your local copy to your remote  branch... I am sorry about this 🤕

Matt-Yorkley commented on spec/features/admin/shipping_methods_spec.rb:
```diff
@@ -88,6 +89,7 @@
       within(".side_menu") do
         click_link "Shipping Methods"
       end
+      DefaultShippingCategory.find_or_create
```
Is this line needed?

daningenthron replied:
Hi @Matt-Yorkley thanks for taking a look. This line was a hacky way to add a second default category to the one already provided by the Spree factory. I've added a new commit that overrides the Spree factory and cleans up the spec.

Matt-Yorkley commented on spec/services/default_shipping_category_spec.rb:
```diff
@@ -0,0 +1,38 @@
+  describe '.create!' do
+    it "names the location 'Default'" do
+      shipping_category = described_class.create!
+      expect(shipping_category.name).to eq('Default')
```
It's not a big deal, but there's lots of unnecessary parentheses in this file. You can use `eq 'Default'` here, for example. I'm surprised`rubocop` is not complaining. I'm not sure what IDE or editor you're using, but many have `rubocop` plugins that add tips for code style, and automated linting.

daningenthron replied:
I have the Rubocop linter added to Sublime 3 - it wasn't catching these, but I'll check the style guide to make sure that my settings are correct. Thanks!

luisramos0 commented on spec/services/default_shipping_category_spec.rb:
```diff
@@ -4,18 +4,18 @@
   describe 'find_or_create' do
     context 'when a Default category already exists' do
-      let!(:category) do
+      let! :category do
```
you need to keep the parenthesis here to keep consistency with all other specs.
in OFN code, not using parenthesis is almost only used in specs after eq

luisramos0 commented on spec/factories.rb:
```diff
@@ -194,4 +194,9 @@
+  factory :shipping_category, class: Spree::ShippingCategory do
+    initialize_with { DefaultShippingCategory.find_or_create }
+    transient { name 'Default' }
+  end
```
:+1:

sigmundpetersen commented:
Semaphore build has 1 error

kristinalim commented on app/views/spree/admin/products/_shipping_category_form.html.haml:
```diff
@@ -1,4 +1,4 @@
 = f.field_container :shipping_categories do 
   = f.label :shipping_category_id, t(:shipping_category)
-  = f.collection_select(:shipping_category_id, Spree::ShippingCategory.all, :id, :name, {:include_blank => true}, {:class => 'select2 fullwidth'})
+  = f.collection_select(:shipping_category_id, Spree::ShippingCategory.all, :id, :name, {:include_blank => false, :selected => DefaultShippingCategory.find_or_create.id}, {:class => 'select2 fullwidth'})
```
This selects "Default" no matter what, which would be a problem in this scenario:

1. In the new product form, fill in required product information except product name.
2. Select a non-default shipping category.
3. Submit the form.
4. In the product form showing the errors, the "Default" shipping category is selected even when it should be a different one.

How about setting the default category for the object in `spree/admin/products#new`? (`app/controllers/spree/admin/products_controller_decorator.rb`). Something like:

```ruby
def new
  @object.shipping_category = DefaultShippingCategory.find_or_create
  super
end
```

Then, this wouldn't be called after the user's unsuccessful submission of the form (`#create`).

daningenthron replied:
Hi @kristinalim - thanks, this definitely makes sense. I had been thinking of a front-end approach with just the forms, instead of the object itself. I see the behavior you're describing in both forms upon submission. This seems to be working now, but I'll retest everything before submitting the PR again.

Regarding the question of the shipping method spec above (adding a second shipping category), I believe this has been resolved with the changes that I made to the shipping category factory.

kristinalim commented on app/views/spree/admin/shipping_methods/_form.html.haml:
```diff
@@ -53,7 +53,7 @@
         - Spree::ShippingCategory.all.each do |category|
           = label_tag do
-            = check_box_tag('shipping_method[shipping_categories][]', category.id, @shipping_method.shipping_categories.include?(category))
+            = check_box_tag('shipping_method[shipping_categories][]', category.id, category == DefaultShippingCategory.find_or_create)
```
Same issue here - the "Default" checkbox is the only one selected when you load the edit shipping method page, even when it has been associated to all available shipping methods.

daningenthron replied:
Thanks again for pointing this out! Since the shipping method accepts multiple categories, I used this variation to instantiate the class:

`def new
        @object.shipping_categories = [DefaultShippingCategory.find_or_create]
        super
end`

kristinalim commented on app/services/default_shipping_category.rb:
```diff
@@ -0,0 +1,13 @@
+# Encapsulates the concept of default stock location in creation of a product or a shipping method.
+
+class DefaultShippingCategory
+  NAME = 'Default'.freeze
```
OFN is deployed in different places around the world, and translations are quite important. For example, there might be some languages where users might not understand what "Default" means.

I checked all [managed instances](https://github.com/openfoodfoundation/ofn-install/wiki/Current-deployment-status). and all still have the "Default" (English) shipping category. But if for example OFN France updates their "Default" to "Défaut", the "Default" category would automatically be created again so they would end up with 2.

This is not an immediate issue that should be fixed, I think. I'll create a separate issue for this.

daningenthron replied:
This is a good point; I'll keep in mind that any user-facing text should be set up with a key in en.yml so Transifex can build out the translation. If you are setting up an issue for this, I modeled `default_shipping_category.rb` after `default_stock_location.rb`, which will need the same fix.

kristinalim replied:
Thanks, @daningenthron!

I created issue #4360 for making the "Default" shipping category translatable.

I didn't create an issue for stock location anymore because we only support one stock location, so this is mostly used internally. :smiley:

daningenthron commented:
@kristinalim Sorry for the delay! Was traveling and wanted to check this against the new controller before pushing. Let me know what you think - thanks!

JenRSmith commented:
Tested in UK and successfully showing default shipping method without requiring user input, and also retaining user selected shipping method rather than overriding with default. 

Testing notes here https://docs.google.com/document/d/1g2yCXRRgKhp3sst08j-qk4LuJ5W6GTgQw34J29rzFFw/edit#

Moved to Ready to Go
